lib/irb/cmd/show_source.rb

```diff
@@ -59,11 +59,18 @@ def find_source(str)
       def find_end(file, first_line)
         return first_line unless File.exist?(file)
         lex = RubyLex.new
-        code = +""
-        File.read(file).lines[(first_line - 1)..-1].each_with_index do |line, i|
-          _ltype, _indent, continue, code_block_open = lex.check_state(code << line)
+        lines = File.read(file).lines[(first_line - 1)..-1]
+        tokens = RubyLex.ripper_lex_without_warning(lines.join)
+        prev_tokens = []
+
+        # chunk with line number
+        tokens.chunk { |tok| tok[0][0] }.each do |lnum, chunk|
+          code = lines[0..lnum].join
```

Comment:
```suggestion
        code = +""
        prev_tokens = []

        # chunk with line number
        tokens.chunk { |tok| tok[0][0] }.each do |lnum, chunk|
          code << lines[lnum]
```

You changed the way `code` is constructed, but this seems like a new bottleneck in your code.

### before
```
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
Rehearsal --------------------------------------------------------
                       2.401019   0.011546   2.412565 (  2.414266)
----------------------------------------------- total: 2.412565sec

                           user     system      total        real
                       2.420137   0.005364   2.425501 (  2.426264)
```

### after
```
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
Rehearsal --------------------------------------------------------
                       1.028879   0.009408   1.038287 (  1.039615)
----------------------------------------------- total: 1.038287sec

                           user     system      total        real
                       1.000221   0.007454   1.007675 (  1.008295)
```

Reply:
(The code I suggested might skip lines that don't have any token, but I assume such lines don't impact compilation error checks.)